Allow emitting debugging symbols via high level scripting APIs

filipw opened the pull request:
**Customer scenario**

As a user of scripting APIs you'd like to be able to consume the symbols to step through the script source. This is not a mainstream scenario, but very useful for script runner authors as they can offer debugging experience.

This was discussed with @tmat and @ManishJayaswal and the suggestion was to open a PR.

**Bugs this fixes:** 

Fixes https://github.com/dotnet/roslyn/issues/12046
Fixes https://github.com/dotnet/roslyn/issues/1543

**Workarounds, if any**

Currently this can be achieved with lots of reflection against the Roslyn APIs and manually emitting the compilation. See example here https://github.com/filipw/dotnet-script/blob/master/src/Dotnet.Script.Core/DebugScriptRunner.cs#L25

However, this is hardly ideal.

**Risk**

Low. Some new public members on `ScriptOptions`

**Performance impact**

Low. The original execution path is unaffected.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

At the moment Scripting APIs completely ignores PDBs and passes `null` into "pdb symbols" when invoking `Compilation.Emit`.

**How was the bug found?**

Lots of community requests for supporting debugging in script runners. 

**Usage & Notes**

```csharp
    class Program
    {
        static void Main(string[] args)
        {
            var path = "foo.csx";
            var txt = File.ReadAllText(path);
            var script = CSharpScript.Create(txt, ScriptOptions.Default.WithDebugInformation(DebugInformationFormat.Pdb).
                WithFilePath(path));
            script.RunAsync().Wait();

           // or just:
           // CSharpScript.RunAsync(txt, ScriptOptions.Default.WithDebugInformation(DebugInformationFormat.Pdb).WithFilePath(path)).Wait();

            Console.ReadKey();
        }
    }
```

The new setting on the `ScriptOptions` allows the user to specify if debug information should be emitted and what `DebugInformationFormat` to use. If so, then when compilation is emitted, symbols are created and loaded.

~~Additionally, the `EmitPdb` setting is used to control `optimizationLevel` on the script compilation.~~ (removed this as it created some regression on the available stacktrace in case of errors)

In order for debugging to work, encoding is mandatory so I also added a possibility to specify encoding via `ScriptOptions` with the default being `UTF8`. 

The new settings (encoding + emit pdb) have not been added to the `ScriptOptions` constructor because the defaults (`UTF8` + `false`) seem reasonable to me.

filipw commented:
The end result:

<img width="400" alt="screenshot 2017-01-13 10 07 10" src="https://cloud.githubusercontent.com/assets/1710369/21924144/08f6cdda-d978-11e6-8a67-2fdf030978e6.png">

tmat commented on src/Scripting/Core/ScriptCompiler.cs:
```diff
@@ -3,6 +3,7 @@
 using System;
 using System.Text;
 using System.Threading;
+using Microsoft.CodeAnalysis.Emit;
```
Unused?

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -8,10 +8,11 @@
 
 namespace Microsoft.CodeAnalysis.Scripting
 {
+    using Microsoft.CodeAnalysis.Emit;
```
>using Microsoft.CodeAnalysis.Emit; [](start = 4, length = 34)

Move outside the namespace

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -92,6 +93,21 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
+        /// <summary>
+        /// <see cref="DebugInformationFormat"/> to be used for debugging symbols if needed
+        /// </summary>
+        public DebugInformationFormat? DebugInformationFormat { get; private set; } = null;
```
Let's remove this option and just use Portable PDB.

tmat replied:
>DebugInformationFormat? [](start = 15, length = 23)

Nullable is not needed, we can use default(DebugInformationFormat). I guess we should add DebugInformationFormat.None to the enum -- it should be there per .NET Design Guidelines.

tmat replied:
>DebugInformationFormat? [](start = 15, length = 23)

Actually, I wonder if we need this option at all. We can detect whether the current platform supports Portable PDBs and only if it doesn't use Windows PDB.

filipw replied:
That sounds like the best idea, I agree. I'm not sure what would be the determining factor though, is it similar to how it is the case for choosing the logic for assembly loading APIs? => if CoreCLR, use PortablePDB, otherwise use full ("properietary") PDB ?

tmat replied:
That's a good question. Ideally we would use Windows PDB only when necessary. On CoreCLR and Mono we would always use Portable PDB. On Desktop FX, if we can detect 4.6.3+ we can use portable as well. 4.6.2 and below doesn't support portable. I'm not sure yet what would the best way to determine this be. Perhaps we can start with CoreCLR and Mono and use Windows PDB on Destkop FX and change that later.

tmat commented on src/Scripting/Core/Hosting/AssemblyLoader/DesktopAssemblyLoaderImpl.cs:
```diff
@@ -30,6 +30,14 @@ public override Assembly LoadFromStream(Stream peStream, Stream pdbStream)
             byte[] peImage = new byte[peStream.Length];
             peStream.TryReadAll(peImage, 0, peImage.Length);
+
+            if (pdbStream != null) {
```
>{ [](start = 35, length = 1)

Style: { on the next line

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -92,6 +93,21 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
+        /// <summary>
+        /// Specifies the encoding to be used for the code.
+        /// </summary>
+        public Encoding Encoding { get; private set; } = Encoding.UTF8;
```
>Encoding [](start = 24, length = 8)

I'd call it FileEncoding and explained in the comment that this is only relevant for debugging scripts that were read from a file or they will be saved to a file for debugging purposes.

tmat commented on src/Scripting/Core/ScriptBuilder.cs:
```diff
@@ -115,20 +115,28 @@ private static void ThrowIfAnyCompilationErrors(DiagnosticBag diagnostics, Diagn
         private Func<object[], Task<T>> Build<T>(
             Compilation compilation,
-            DiagnosticBag diagnostics,
+            DiagnosticBag diagnostics, ScriptOptions scriptOptions,
```
> ScriptOptions scriptOptions, [](start = 38, length = 29)

Style: on a new line.

tmat replied:
I'd just pass DebugInformationFormat to this method because that's the only thing needed here.

---
In reply to: [96129988](https://github.com/dotnet/roslyn/pull/16489#discussion_r96129988) [](ancestors = 96129988)

tmat commented on src/Scripting/Core/ScriptBuilder.cs:
```diff
@@ -115,20 +115,28 @@ private static void ThrowIfAnyCompilationErrors(DiagnosticBag diagnostics, Diagn
             var entryPoint = compilation.GetEntryPoint(cancellationToken);
 
             using (var peStream = new MemoryStream())
+            using (var pdbStream = scriptOptions.EmitDebugInformation ? new MemoryStream() : null)
```
>pdbStream [](start = 23, length = 9)

Let's call this variable pdbStreamOpt as it can be null

tmat commented:
Great start, thanks for taking initiative on this! I have commented on the changes. In addition I'd like to see more tests that validate the presence of debug information. I think we could use StackTrace API to see whether there is debug info or not for the compiled script.

filipw commented:
Thanks a lot for the review - will apply the changes 👍

filipw commented:
@tmat 

I've applied the code changes requested and simplified the API - as you suggested, instead of passing in `DebugInformationFormat` into `ScriptOptions` we now expose just `ScriptOptions.Default.WithDebugInformation()`. Then we internally infer the PDB format -> PortablePdb for CoreCLR/Mono and "classic" for Windows desktop.  

I also added unit tests to verify the presence of PDB (check for line number in the stack trace).

PS. I also noticed this PR, aside from the 2 originally mentioned issues, also fixes https://github.com/dotnet/roslyn/issues/13482

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -282,5 +295,17 @@ private static MetadataReference CreateReferenceFromAssembly(Assembly assembly)
+        /// <summary>
+        /// Creates a new <see cref="ScriptOptions"/> with debugging information enabled.
+        /// </summary>
+        public ScriptOptions WithDebugInformation() =>
```
>WithDebugInformation [](start = 29, length = 20)

To follow the With- pattern this should be ```WithEmitDebugInformation(bool value)```

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -123,6 +134,8 @@ private ScriptOptions(ScriptOptions other)
                    sourceResolver: other.SourceResolver)
         {
+            EmitDebugInformation = other.EmitDebugInformation;
+            FileEncoding = other.FileEncoding;
```
Let's add corresponding parameters to the constructor.

filipw replied:
👍

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -92,6 +93,16 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
+        /// Specifies the encoding to be used when debugging scripts loaded from a file, or that will be saved to a file for debugging purposes.
+        /// </summary>
+        public Encoding FileEncoding { get; private set; } = Encoding.UTF8;
+
```
Having default encoding might cause confusion. Even if the file uses UTF8 encoding there are still two kinds of encoding - with BOM and without. If it differs the script won't be debuggable. I'd rather require the user to explicitly specify encoding than to run into these subtleties with the default value.

filipw replied:
ok will change to `null` (that would be part of `ScriptOptions.Default` too)

daveaglick commented:
@filipw Been watching this PR with interest - thanks for doing the work to put it together, and thanks to @tmat for the quick review

tmat commented on src/Scripting/CSharpTest/ScriptTests.cs:
```diff
@@ -687,6 +687,35 @@ public void TestCreateMethodDelegate()
+            catch (Exception ex)
+            {
+                // line information is only available when PDBs have been emitted
+                Assert.DoesNotContain("at Submission#0.<<Initialize>>d__0.MoveNext() in :line 1", ex.StackTrace);
```
>:line 1 [](start = 88, length = 7)

This is odd, why do we have line number info here?

It would also be good to set the FileName and validate it's included in the stack frame string.

filipw replied:
this was intended to be a negative test (does not contain the line number info). I will change it to a specific stack trace without line number, it will be much more explicit

tmat commented on src/Scripting/Core/ScriptBuilder.cs:
```diff
@@ -66,7 +66,7 @@ public int GenerateSubmissionId(out string assemblyName, out string typeName)
 
         /// <exception cref="CompilationErrorException">Compilation has errors.</exception>
-        internal Func<object[], Task<T>> CreateExecutor<T>(ScriptCompiler compiler, Compilation compilation, CancellationToken cancellationToken)
+        internal Func<object[], Task<T>> CreateExecutor<T>(ScriptCompiler compiler, Compilation compilation, ScriptOptions scriptOptions, CancellationToken cancellationToken)
```
>ScriptOptions scriptOption [](start = 109, length = 26)

Let's just pass ```bool emitDebugInformation```

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -282,5 +295,17 @@ private static MetadataReference CreateReferenceFromAssembly(Assembly assembly)
+        /// Creates a new <see cref="ScriptOptions"/> with specified <see cref="FileEncoding"/>.
+        /// </summary>
+        public ScriptOptions WithFileEncoding(Encoding encoding) =>
+            new ScriptOptions(this) { FileEncoding = encoding };
```
>new ScriptOptions(this) [](start = 12, length = 23)

Check if FileEncoding == encoding to avoid creating new instance if the values are equal.

tmat commented on src/Scripting/CSharpTest/ScriptTests.cs:
```diff
@@ -687,6 +687,35 @@ public void TestCreateMethodDelegate()
+            {
+                // line information is only available when PDBs have been emitted
+                Assert.DoesNotContain("at Submission#0.<<Initialize>>d__0.MoveNext() in :line 1", ex.StackTrace);
+            }
```
Add a test where FileEncoding is null.

filipw replied:
👍

filipw commented:
@tmat thanks for quick responses, much appreciated

tmat commented on src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs:
```diff
@@ -149,7 +149,9 @@ private static ScriptOptions GetScriptOptions(CommandLineArguments arguments, st
-                sourceResolver: sourceResolver);
+                sourceResolver: sourceResolver,
+                emitDebugInformation: false,
+                fileEncoding: null);
```
Doesn't need to be fixed now but shouldn't the command line runner (csi) actually use this feature, so that we have better exception stack traces?

filipw replied:
yes that's what I thought as well - let's add it in a separate PR

filipw commented:
@tmat can you take a look again? applied the feedback + improved the tests

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -282,5 +301,17 @@ private static MetadataReference CreateReferenceFromAssembly(Assembly assembly)
+        /// <summary>
+        /// Creates a new <see cref="ScriptOptions"/> with specified <see cref="FileEncoding"/>.
+        /// </summary>
+        public ScriptOptions WithFileEncoding(Encoding encoding) => 
```
>WithFileEncoding [](start = 29, length = 16)

Just realized passing the encoding in options is not enough. The problem is that encodings are not 1:1 mappings in general. Say you have a script file encoded by encoding E saved on disk.  You read it to a string using decoder for E and then create a Script from that string and also pass in encoding E. Then to calculate the checksum for the source file that's stored in the PDB, we need to convert the string back to bytes using encoder for E. The problem is that for some encodings the resulting bytes might not be exactly the same that the file contains, unless the encoding is 1:1 mapping. 

Also, the compiler implements somewhat sophisticated encoding detection. So it would be unfortunate if the user needed to reimplement that. 

I'd suggest that we add overloads to core Scripting methods that take ```string code``` today. We don't need to add overloads to all methods since most of them are just convenience APIs that call the core ones.

I'm thinking these 3:

```C#
CSharpScript.Create(Stream code, ...)
Script.ContinueWith(Stream code, ...)
Script.ContinueWith<T>(Stream code, ...)
```

Then internally within Scripting API implementation we'd create SourceText using From(Stream, Encoding) method.

Could you give it a try?

filipw replied:
sure, so the suggestion - if I understand correctly - is to rely on letting encoding be detected by `SourceText.From(stream)` because internally it would try to do it by looking at BOM etc? 

When we add the stream overloads, do you also remove manual encoding setting from options and specify that using the non-stream scripting methods will not allow debugging (since under the hood it will use non-debuggable `SourceText`)?

tmat replied:
No, we keep the ```FileEncoding``` property.  If the encoding is not passed to SourceText the detection is attempted (using BOM) and if we can't detect we default to UTF8 with no BOM. If there is no BOM the encoding needs to be specified, which is what FileEncoding would be used for. You can look at the implementation of http://source.roslyn.io/#Microsoft.CodeAnalysis/Text/SourceText.cs,3687bd0c8922a81f to see what it's doing. Note the exceptions it may throw. We should pass them on the to the caller of the script APIs that take Stream and document them.

filipw replied:
OK sounds good. 

One thing I noticed straight away though is that adding the overloaded methods with `Stream`, will cause the analyzer to complain: `Symbol Create<T> violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'`

tmat replied:
What analyzer is that?

tmat replied:
That seems like a bug in the analyzer.

filipw replied:
Looks like something from @jaredpar 

```
error RS0026: Symbol 'Create<T>' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'. See 'https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md' for details.
```

tmat replied:
There are multiple suppressions of this error already. Just suppress it as well for now. We should fix the analyzer.

filipw replied:
👍

jaredpar replied:
This does not line up with how we've been approaching our APIs.  There should be a single overload with optional parameters.  That is part of the versioning story. 

@Pilchie

jaredpar replied:
> There are multiple suppressions of this error already

That's not good.  When did these pop up?  Never went over any of these during API reviews.  If this is happening we need to start flagging them and reviewing.

tmat replied:
The rule is overly conservative. We can version these overloads methods just fine.

tmat replied:
I filed an issue here: https://github.com/dotnet/roslyn/issues/16564

jaredpar replied:
> The rule is overly conservative. We can version these overloads methods just fine.

That may be true.  It's also not a decision we can not be making in a PR in a release branch during lock down.

filipw commented:
I pushed some changes - I'm not sure how much tests there should be, given that it all ends up calling into the same underlying stuff, the stream methods are merely shortcuts.

tmat commented on src/Scripting/Core/Script.cs:
```diff
@@ -86,12 +98,30 @@ internal static Script<T> CreateInitialScript<T>(ScriptCompiler compiler, string
         public Script<object> ContinueWith(string code, ScriptOptions options = null) =>
             ContinueWith<object>(code, options);
 
+        #pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
```
>#pragma warning disable RS0026 // Do not add multiple p [](start = 8, length = 55)

You can just move this to the top of the file, no need to repeat multiple times.

tmat commented on src/Scripting/Core/Script.cs:
```diff
@@ -279,15 +309,25 @@ public sealed class Script<T> : Script
 
         internal Script(ScriptCompiler compiler, ScriptBuilder builder, string code, ScriptOptions options, Type globalsTypeOpt, Script previousOpt)
-            : base(compiler, builder, code, options, globalsTypeOpt, previousOpt)
+            : base(compiler, builder, SourceText.From(code, options.FileEncoding), options, globalsTypeOpt, previousOpt)
```
> SourceText.From(code, options.FileEncoding) [](start = 37, length = 44)

Can we move this up the stack and keep just one ctor that takes SourceText?

tmat commented:
>         internal static Script<T> CreateInitialScript<T>(ScriptCompiler compiler, string codeOpt, ScriptOptions optionsOpt, Type globalsTypeOpt, InteractiveAssemblyLoader assemblyLoaderOpt)

I'd change this to take SourceText to avoid duplication of the logic in the body.

---
Refers to: src/Scripting/Core/Script.cs:47 in 415e7a7. [](commit_id = 415e7a7c3b5750e832cdc7b364216b8c1e09a838, deletion_comment = False)

tmat commented on src/Scripting/CSharp/CSharpScript.cs:
```diff
@@ -26,6 +27,22 @@ public static Script<T> Create<T>(string code, ScriptOptions options = null, Typ
             return Script.CreateInitialScript<T>(CSharpScriptCompiler.Instance, code, options, globalsType, assemblyLoader);
         }
 
+        #pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
```
>#pragma w [](start = 8, length = 9)

Move to the top of the file.

tmat commented on src/Scripting/Core/Script.cs:
```diff
@@ -62,7 +69,12 @@ internal static Script<T> CreateInitialScript<T>(ScriptCompiler compiler, string
+        /// <summary>
+        /// The <see cref="SourceText"/> of the script.
+        /// </summary>
+        public SourceText SourceText { get; }
```
>public SourceText SourceText { get; } [](start = 8, length = 37)

Let's make this internal. I don't think it needs to be exposed right now.

filipw commented:
looks like one of the builds didn't report back but it completed successfully too

filipw commented:
I pushed some extra tests covering the PDB scenario via the new entry path (`CSharpScript.Create(stream)`)

jaredpar commented on src/Scripting/CSharp/CSharpScript.cs:
```diff
@@ -1,10 +1,13 @@
 // Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
```
We should not be disabling this analyzer.  Need to address the problem and settle on a single overload with optional parameters.

tmat replied:
Yes, we need to address the problem in the analyzer. I don't think there is anything wrong with the API.

filipw commented:
Should also fix #13650 (see comment https://github.com/dotnet/roslyn/issues/13650#issuecomment-251239726)

tmat commented:
@filipw Sorry for the delay. I've been busy with RTM issues. Back now. We need to move this PR to post-dev15 branch. Could you please rebase?

tmat commented on src/Scripting/Core/ScriptOptions.cs:
```diff
@@ -92,6 +95,16 @@ private static ImmutableArray<MetadataReference> GetDefaultMetadataReferences()
+        public bool EmitDebugInformation { get; private set; } = false;
+
+        /// <summary>
+        /// Specifies the encoding to be used when debugging scripts loaded from a file, or that will be saved to a file for debugging purposes.
```
>Specifies  [](start = 12, length = 10)

Document that the value may be null.

tmat commented on src/Scripting/Core/Script.cs:
```diff
@@ -87,10 +95,26 @@ internal static Script<T> CreateInitialScript<T>(ScriptCompiler compiler, string
         /// </summary>
         public Script<TResult> ContinueWith<TResult>(string code, ScriptOptions options = null) =>
-            new Script<TResult>(Compiler, Builder, code ?? "", options ?? InheritOptions(Options), GlobalsType, this);
+            new Script<TResult>(Compiler, Builder, SourceText.From(code ?? "", options?.FileEncoding ?? Options.FileEncoding), options ?? InheritOptions(Options), GlobalsType, this);
```
>options?.FileEncoding [](start = 79, length = 21)

I don't think we should use ?? here -- if the FileEncoding is set to null on the current options, we shouldn't use the previous script encoding. 

Let's simplify this complex expression a bit to something like:
```
{
options = options ?? InheritOptions(Options);
return new Script(..., options.FileEncoding), options, ...)
}
```

tmat commented on src/Scripting/Core/Script.cs:
```diff
@@ -87,10 +95,26 @@ internal static Script<T> CreateInitialScript<T>(ScriptCompiler compiler, string
+        /// <exception cref="ArgumentNullException">Stream is null.</exception>
+        /// <exception cref="ArgumentException">Stream is not readable or seekable.</exception>
+        public Script<TResult> ContinueWith<TResult>(Stream code, ScriptOptions options = null) =>
+            new Script<TResult>(Compiler, Builder, SourceText.From(code, options?.FileEncoding ?? Options.FileEncoding), options ?? InheritOptions(Options), GlobalsType, this);
```
>options?.FileEncoding ?? Options.FileEncodin [](start = 73, length = 44)

Same as above

tmat replied:
>code [](start = 67, length = 4)

check for code == null. Just a detail, but we should be throwing exceptions with the correct parameter name.

tmat commented on src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs:
```diff
@@ -176,7 +178,7 @@ private int RunScript(ScriptOptions options, string code, ErrorLogger errorLogge
             globals.Args.AddRange(_compiler.Arguments.ScriptArguments);
 
-            var script = Script.CreateInitialScript<int>(_scriptCompiler, code, options, globals.GetType(), assemblyLoaderOpt: null);
+            var script = Script.CreateInitialScript<int>(_scriptCompiler, SourceText.From(code ?? string.Empty), options, globals.GetType(), assemblyLoaderOpt: null);
```
>code ?? [](start = 90, length = 7)

I believe code can't be null here

tmat commented on src/Scripting/Core/Hosting/CommandLine/CommandLineRunner.cs:
```diff
@@ -249,7 +251,7 @@ private void RunInteractiveLoop(ScriptOptions options, string initialScriptCodeO
                 if (state == null)
                 {
-                    newScript = Script.CreateInitialScript<object>(_scriptCompiler, code, options, globals.GetType(), assemblyLoaderOpt: null);
+                    newScript = Script.CreateInitialScript<object>(_scriptCompiler, SourceText.From(code ?? string.Empty), options, globals.GetType(), assemblyLoaderOpt: null);
```
>code [](start = 100, length = 4)

code can't be null here

tmat commented on src/Scripting/CoreTest/ScriptOptionsTests.cs:
```diff
@@ -148,5 +150,35 @@ public void WithImports_Errors()
+
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
```
nit: this is a bit of overkill.

Assert.True(ScriptOptions.Default.WithEmitDebugInformation(true));
Assert.False(ScriptOptions.Default.WithEmitDebugInformation(false));

would do.

tmat commented on src/Scripting/CoreTest/ScriptOptionsTests.cs:
```diff
@@ -148,5 +150,35 @@ public void WithImports_Errors()
+        [Fact]
+        public void WithFileEncoding_SetsWithFileEncoding()
+        {
+            var options = ScriptOptions.Default.WithFileEncoding(Encoding.ASCII);
```
>Encoding.ASCII) [](start = 65, length = 15)

Missing case: null

tmat commented on src/Scripting/CSharp/CSharpScript.cs:
```diff
@@ -23,7 +26,22 @@ public static class CSharpScript
+        /// <exception cref="ArgumentNullException">Stream is null.</exception>
+        /// <exception cref="ArgumentException">Stream is not readable or seekable.</exception>
+        public static Script<T> Create<T>(Stream code, ScriptOptions options = null, Type globalsType = null, InteractiveAssemblyLoader assemblyLoader = null)
+        {
```
>{ [](start = 8, length = 1)

check code == null and throw

tmat commented on src/Scripting/CSharpTest/ScriptTests.cs:
```diff
@@ -687,6 +712,97 @@ public void TestCreateMethodDelegate()
+                var script = await CSharpScript.RunAsync("throw new System.Exception();", opts);
+            }
+            catch (CompilationErrorException ex)
+            {
```
>{ [](start = 12, length = 1)

Use Verify helper like so:

parsedArgs.Errors.Verify(Diagnostic(ErrorCode....))

tmat commented on src/Scripting/CSharpTest/ScriptTests.cs:
```diff
@@ -739,5 +855,32 @@ public override Stream OpenRead(string resolvedPath)
+            }
+
+            return null;
+        }
```
Rather than parsing the string it would be better to use ```new StackTrace(exception, needFileInfo: true)``` APIs.

filipw commented:
@tmat thanks a lot! 
I rebased on top of `post-dev15` and applied the requested changes. Let me know if something still stands out

tmat commented:
It seems that github is not doing the rebase right. The PR now has 300+ commits :(

@jasonmalinowski @tannergooding

filipw commented:
should we repoint the PR to the post-dev15 branch? at the moment we are targeting a merge to master

jaredpar commented:
The rebase looks correct.  Problem is the PR is still targeting master as the base.  It needs to move to post-dev15 as well.

tannergooding commented:
Modified the target branch to be correct. Back down to a `+370 −30` change

tmat commented on src/Scripting/CSharp/CSharpScript.cs:
```diff
@@ -23,7 +26,23 @@ public static class CSharpScript
         /// <typeparam name="T">The return type of the script</typeparam>
         public static Script<T> Create<T>(string code, ScriptOptions options = null, Type globalsType = null, InteractiveAssemblyLoader assemblyLoader = null)
         {
-            return Script.CreateInitialScript<T>(CSharpScriptCompiler.Instance, code, options, globalsType, assemblyLoader);
```
>return [](start = 12, length = 6)

throw if code == null here as well

tmat replied:
and add cod comment that this method throws ArgNullEx


---
In reply to: [99941131](https://github.com/dotnet/roslyn/pull/16489#discussion_r99941131) [](ancestors = 99941131)

tmat commented:
:shipit:

tmat commented:
Looks good, other than one last minor thing :) 
I'm gonna merge as soon as it's addressed.

filipw commented:
thanks - should be all good now 🙈

tmat commented:
@filipw Thanks a lot for making it happen!

filipw commented:
let's enable this for CSI out of the box 😄
